std/range/primitives.d

```diff
@@ -161,16 +161,19 @@ See_Also:
 
 Params:
     R = type to be tested
+    E = the element type of the range. `void` means "any". Applies Unqual to E.
 
 Returns:
     `true` if R is an input range, `false` if not
  */
-enum bool isInputRange(R) =
+enum bool isInputRange(R, E = void) =
     is(typeof(R.init) == R)
     && is(ReturnType!((R r) => r.empty) == bool)
     && is(typeof((return ref R r) => r.front))
     && !is(ReturnType!((R r) => r.front) == void)
-    && is(typeof((R r) => r.popFront));
+    && is(typeof((R r) => r.popFront))
+    && (is(E == void) || is(Unqual!(ElementType!R) == E));
```

Comment:
Why `is(Unqual!(ElementType!R) == E))` instead of `is(R : E)` (like in #3786)? Is there an example of where this way does the right thing and the other way does the wrong thing?

Reply:
Implicit conversions are generally a terrible idea in generic code. At minimum, if a template constraint accepts an implicit conversion, then it needs to force the implicit conversion, or you risk bugs - some of which result in compile-time errors, and some of which just result in weird, buggy behavior (e.g. if some of the function uses the object as-is, and other parts trigger the implicit conversion, which can happen easily if some of the stuff done to the object works without the conversion and other stuff only works if the conversion is done). So, having a commonly-used trait use `:` instead of `==` is just begging for problems. Anyone using `:` needs to know that that's what they're doing and program accordingly and _not_ have that be the default.

Using `Unqual` is already potentially risky. It's needed so often that it's probably okay, but it's not hard to write code that works with mutable and not other levels of constness such that just letting `const` and `immutable` elements through will cause problems some of the time - though at least in those cases, it's virtually guaranteed to be a compilation error. So, the risk of problems is probably outweighed by the usability gained by then working in all of the cases where you don't care about constness.

But `:` should be used in template constraints with extreme caution.